### PR TITLE
Formalised Erdos 859

### LINKED ISSUE #984: Erdős Problem 859
### What is the conjecture

https://www.erdosproblems.com/859

  Let $t\geq 1$ and let $d_t$ be the density of the set of integers $n\in\mathbb{N}$ for which $t$ can be represented as the sum of distinct divisors of $n$.

Do there exist constants $c_1,c_2>0$ such that
$$d_t \sim \frac{c_1}{(\log t)^{c_2}}$$
as $t\to \infty$?

Status:  open

### Choose either option
- [ ] I plan on working on this conjecture
- [x] This issue is up for grabs: I would like to see this conjecture added by somebody else

### PR DESCRIPTION
**Pull request by KaJabben**
I added a formalization of Erdos 859.

fixes #984

### CODE REVIEW

File: FormalConjectures/ErdosProblems/859.lean [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+*Reference:* [erdosproblems.com/859](https://www.erdosproblems.com/859)
+-/
+
+namespace Erdos859
+
+/--
+`DivisorSumSet t` is the set of natural numbers `n` such that `t` can be represented as
+a sum of divisors of `n`.
+-/
+def DivisorSumSet (t : ℕ) := { n : ℕ | ∃ s ⊆ (Nat.divisors n), t = ∑ i ∈ s , i }
```
Comment (callesonne) [inline FormalConjectures/ErdosProblems/859.lean]:
```suggestion
def DivisorSumSet (t : ℕ) := { n : ℕ | ∃ s ⊆ Nat.divisors n, t = ∑ i ∈ s , i }
```
I think no parenthesis is needed here (but correct me if I'm wrong)

File: FormalConjectures/ErdosProblems/859.lean [hunk 2]
```diff
@@ -0,0 +1,67 @@
...
+
+open Asymptotics
+open Filter
```
Comment (callesonne) [inline FormalConjectures/ErdosProblems/859.lean]:
```suggestion
open Asymptotics Filter
```

File: FormalConjectures/ErdosProblems/859.lean [hunk 3]
```diff
@@ -0,0 +1,67 @@
...
+# Erdős Problem 859
+
+*Reference:* [erdosproblems.com/859](https://www.erdosproblems.com/859)
+-/
+
+namespace Erdos859
+
+/--
+`DivisorSumSet t` is the set of natural numbers `n` such that `t` can be represented as
+a sum of divisors of `n`.
```
Comment (callesonne) [inline FormalConjectures/ErdosProblems/859.lean]:
```suggestion
a sum of distinct divisors of `n`.
```

File: FormalConjectures/ErdosProblems/859.lean [hunk 4]
```diff
@@ -0,0 +1,67 @@
...
+
+open Asymptotics
+open Filter
+
+/-- A weaker version of the problem proved by Erdos:
+The density `dₜ` of `DivisorSumSet (t : ℕ)` is bounded from below by `1 / log (t) ^ c₃` and
+from above by `1 / log (t) ^ c₄` for some positive constants `c₃` and `c₄`.
+-/
+@[category research solved, AMS 11]
+theorem erdos_859.variants.erdos_upper_lower_bounds : ∃ c₃ > 0, ∃ c₄ > 0, ∃ t₀ : ℕ, ∀ t > t₀,
```
Comment (callesonne) [inline FormalConjectures/ErdosProblems/859.lean]:
```suggestion
theorem erdos_859.variants.erdos_upper_lower_bounds : ∃ᵉ (c₃ > 0) (c₄ > 0) (t₀ : ℕ), ∀ t > t₀,
```
This makes it a bit more compact :)

File: FormalConjectures/ErdosProblems/859.lean [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+
+/-
+**Erdős Problem 859**
+The density `dₜ` of `DivisorSumSet (t : ℕ)` is assymptotically equivalent to ` c₁ / log (t) ^ c₂`
+for some positive constants `c₁` and `c₂`.
+-/
+@[category research open, AMS 11]
+theorem erdos_859 (t : ℕ) : ∃ c₁ > 0, ∃ c₂ > 0, ∃ dₜ : ℝ, ((DivisorSumSet t).HasDensity dₜ) ∧
+  (IsEquivalent (atTop : Filter ℕ)  (fun t ↦ dₜ) (fun t ↦ c₁ / (Real.log t)^c₂ )) := by
```
Comment (callesonne) [inline FormalConjectures/ErdosProblems/859.lean]:
```suggestion
  (fun (t : ℕ) ↦ dₜ) ~[atTop] (fun t ↦ c₁ / (Real.log t)^c₂ ) := by
```
You should be able to write it like this in the Asymptotics namespace. I have not checked this locally so it might need small adjustments to compile.

File: FormalConjectures/ErdosProblems/859.lean [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+  sorry
+
+/-
+An easy sanity check is to prove that for every natural number `t` the density `dₜ` is
+a positve number.
+Hint: investigate some multiplicative structure of `DivisorSumSet t`.
+-/
+@[category undergraduate, AMS 11]
+lemma erdos_859.variants.positive_density (t : ℕ) :
+  ∃ dₜ : ℝ, ((DivisorSumSet t).HasDensity dₜ) ∧ dₜ > 0 := by
```
Comment (callesonne) [inline FormalConjectures/ErdosProblems/859.lean]:
```suggestion
  (DivisorSumSet t).HasPosDensity := by
```

**Comment by KaJabben** [comment]
Thanks for the suggestions! I added all of the suggested changes and checked that it still compiles.

File: FormalConjectures/ErdosProblems/859.lean [hunk 7]
```diff
@@ -0,0 +1,66 @@
...
+
+
+/-
+**Erdős Problem 859**
+The density `dₜ` of `DivisorSumSet (t : ℕ)` is assymptotically equivalent to ` c₁ / log (t) ^ c₂`
+for some positive constants `c₁` and `c₂`.
+-/
+@[category research open, AMS 11]
+theorem erdos_859 (t : ℕ) : ∃ c₁ > 0, ∃ c₂ > 0, ∃ dₜ : ℝ, ((DivisorSumSet t).HasDensity dₜ) ∧
+  (fun (t : ℕ) ↦ dₜ) ~[atTop] (fun t ↦ c₁ / (Real.log t)^c₂ ) := by
```
Comment (Paul-Lez) [inline FormalConjectures/ErdosProblems/859.lean]:
```suggestion
    (fun (t : ℕ) ↦ dₜ) ~[atTop] (fun t ↦ c₁ / (Real.log t)^c₂ ) := by
```

File: FormalConjectures/ErdosProblems/859.lean [hunk 8]
```diff
@@ -0,0 +1,66 @@
...
+open Asymptotics Filter
+
+/-- A weaker version of the problem proved by Erdos:
+The density `dₜ` of `DivisorSumSet (t : ℕ)` is bounded from below by `1 / log (t) ^ c₃` and
+from above by `1 / log (t) ^ c₄` for some positive constants `c₃` and `c₄`.
+-/
+@[category research solved, AMS 11]
+theorem erdos_859.variants.erdos_upper_lower_bounds : ∃ᵉ (c₃ > 0) (c₄ > 0) (t₀ : ℕ), ∀ t > t₀,
+  ∃ dₜ : ℝ, ((DivisorSumSet t).HasDensity dₜ) ∧
+  (1 / (Real.log t)^c₃ < dₜ) ∧ (dₜ < 1 / (Real.log t)^c₄) := by
```
Comment (Paul-Lez) [inline FormalConjectures/ErdosProblems/859.lean]:
```suggestion
    ∃ dₜ : ℝ, ((DivisorSumSet t).HasDensity dₜ) ∧
    (1 / (Real.log t)^c₃ < dₜ) ∧ (dₜ < 1 / (Real.log t)^c₄) := by
```

File: FormalConjectures/ErdosProblems/859.lean [hunk 9]
```diff
@@ -0,0 +1,66 @@
...
+  sorry
+
+/-
+An easy sanity check is to prove that for every natural number `t` the density `dₜ` is
+a positve number.
+Hint: investigate some multiplicative structure of `DivisorSumSet t`.
+-/
+@[category undergraduate, AMS 11]
+lemma erdos_859.variants.positive_density (t : ℕ) :
+  (DivisorSumSet t).HasPosDensity := by
```
Comment (Paul-Lez) [inline FormalConjectures/ErdosProblems/859.lean]:
```suggestion
    (DivisorSumSet t).HasPosDensity := by
```

File: FormalConjectures/ErdosProblems/859.lean [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+*Reference:* [erdosproblems.com/859](https://www.erdosproblems.com/859)
+-/
+
+namespace Erdos859
+
+/--
+`DivisorSumSet t` is the set of natural numbers `n` such that `t` can be represented as
+a sum of distinct divisors of `n`.
+-/
+def DivisorSumSet (t : ℕ) := { n : ℕ | ∃ s ⊆ Nat.divisors n, t = ∑ i ∈ s , i }
```
Comment (Paul-Lez) [inline FormalConjectures/ErdosProblems/859.lean]:
```suggestion
def DivisorSumSet (t : ℕ) := { n : ℕ | ∃ s ⊆ Nat.divisors n, t = ∑ i ∈ s, i }
```

File: FormalConjectures/ErdosProblems/859.lean [hunk 11]
```diff
@@ -0,0 +1,66 @@
...
+/-
+An easy sanity check is to prove that for every natural number `t` the density `dₜ` is
+a positve number.
+Hint: investigate some multiplicative structure of `DivisorSumSet t`.
+-/
+@[category undergraduate, AMS 11]
+lemma erdos_859.variants.positive_density (t : ℕ) :
+  (DivisorSumSet t).HasPosDensity := by
+  sorry
+
```
Comment (Paul-Lez) [inline FormalConjectures/ErdosProblems/859.lean:72]: I think it could be worth adding an extra `test` statement that shows that `DivisorSumSet 0 = Set.univ` to document this edge case.

File: FormalConjectures/ErdosProblems/859.lean [hunk 12]
```diff
@@ -0,0 +1,73 @@
...
+/-
+A case where we can easily calculate the density of `DivisorSumSet t` is that of `t=0`.
+-/
+@[category high_school, AMS 11]
+lemma erdos_859.variants.trivial_case : DivisorSumSet 0 = Set.univ := by
+  sorry
+
+/-
+An easy sanity check is to prove that for every natural number `t` the density `dₜ` is
+a positve number.
```
Comment (YaelDillies) [inline FormalConjectures/ErdosProblems/859.lean]:
```suggestion
a positive number.
```
